File: 2300. Successful Pairs of Spells and Potions/Solution.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success)
    {
        sort(potions.begin(), potions.end());
        vector<int> res;
        for (int spell : spells) {
            long long need = (success - 1) / spell;
            res.emplace_back(potions.end() - upper_bound(potions.begin(), potions.end(), need));
        }
        return res;
    }
};

File: 2300. Successful Pairs of Spells and Potions/Solution.py
import bisect


class Solution:
    def successfulPairs(self, spells: list[int], potions: list[int], success: int) -> list[int]:
        potions.sort()
        return [len(potions) - bisect.bisect_right(potions, (success - 1) // spell) for spell in spells]
